make PumpLL idempotent to reduce cpu use a bit

majestrate opened the pull request:
in order to reduce excess cpu use in PathContext's `Pump{Up,Down}stream()` make the only place in the code that calls those functions idempotent (`Router::PumpLL`).

majestrate committed:
make PumpLL idempotent to reduce cpu use a bit

majestrate committed:
prevent 100% cpu usage on service nodes

call_soon wakes up the mainloop and is often reentrant, we dont want to
ever way up the event loop but we also want to always defer the call so
we always use the workers

majestrate committed:
make event loop pump function non idempotent

majestrate committed:
only pump path context in idempotent pumpll

jagerman committed:
Allow a nullptr callback for event loop PumpLL

jagerman committed:
Make outbound message queue PumpLL again if it doesn't send all

jagerman committed:
outbound_message_handler: skip instead of aborting on removed paths

jagerman committed:
transit hop: drop instead of flushing when full

If full happens we are going to trigger a bunch of flushes which
probably isn't very useful, so drop instead.

jagerman committed:
Simplifications & C++17

- Modernize some iterator loops
- Simplify a couple places by using `if (init; ...)`
- Replace various std::binds with lambdas

jagerman committed:
Triggered pumping

This redoes how/when we pump so that we should only be calling the
idempotent PumpLL() when we actually have (or may have) something to
pump.

jagerman committed:
Remove another unwanted unconditional PumpLL

jagerman committed:
Combine empty + pop into single call

jagerman committed:
Fix link layer delivery race condition (fix random ping delays)

We trigger a pump immediately, but this is racey because we add to our
plaintext data in a worker thread; if the worker thread runs after the
pump then it ends up leaving plaintext to be handled, but there's no
wakeup until the next one.

This was the cause of seeing a random +1s and bunching added to ping
responses sometimes: it wasn't until the *next* ping goes through the
network that the plaintext queue gets processed, at which point it
flushes the old one and often the new one together.

The fix here gets rid of the map of sessions needing wakeups and instead
adds an atomic flag to all of them to let us figure out which ones
need to be flushed.

jagerman committed:
Simplify

jagerman committed:
Get rid of external event loop direct wakeups

If something needs to wake up the event loop it should be using an
async, as we are now with PumpLL(); but we had various code triggering a
wakeup, expecting that PumpLL gets called on every wakeup, which isn't
true anymore.

jagerman committed:
Rename PumpLL -> TriggerPump

And rename the actual pump implementation back to PumpLL.

jagerman committed:
Increase plaintext queue size from 32 to 512

In testing we were sometimes hitting a full queue error, which makes
sense because the queue size seems really small.

Increase it to 512.

jagerman committed:
Cleanups

jagerman committed:
Fix packet delay creep

- Make the main PumpLL also pump hidden services, rather than using
  separate wakers in each TunEndpoint.  It seems there is some
  interactions that just one or the other is not enough.

- Eliminate TunEndpoint send queue -- it isn't needed as we can just
  send directly.

jagerman committed:
Cleanup: avoid pointless copy of `this` pointer

jagerman committed:
Remove dead declaration

jagerman committed:
Pump hidden context sooner

This has to happen before pumping links because it is putting things on
links.

jagerman committed:
Simplify

jagerman committed:
Add missing inbound traffic pump trigger

jagerman committed:
Pump outbound context

jagerman committed:
Add another missing pump trigger

jagerman committed:
Fix client latency bug; De-shared_ptr IHopHandler queues

- Replace m_FlushWakeup with a call to the router's god mode pump
  method.  m_FlushWakeup apparently isn't enough to get things out, and
  we can end up with incoming packets that don't get properly handled
  right away without it.

- The shared_ptr around the ihophandler queues isn't needed and is just
  adding a layer of obfuscation; instead just exchange the list directly
  into the lambda.

- Use std::exchange rather than swap

- A couple other small code cleanups.

jagerman committed:
Remove hopeless flush

The TriggerPump just below this is *already* going to trigger a flush,
so the extra flush call here can't do anything useful (and in
particular, it won't clear up the queue *immediately*, which is what
this code looks like it was aimed at doing).

majestrate commented on llarp/handlers/exit.cpp:
```diff
@@ -837,18 +809,14 @@ namespace llarp
         if (itr->second->GetCurrentPath() == ep->GetCurrentPath())
         {
-          itr = m_ActiveExits.erase(itr);
+          m_ActiveExits.erase(itr);
```
this erase will change the thing we are iterating over if it rehashes

jagerman replied:
Yes, but as the next line says right before it returns, "// now ep is gone af"

jagerman committed:
Remove broken test

This test isn't very useful, but also broken with the other changes on
this branch.

tewinget commented on llarp/handlers/exit.cpp:
```diff
@@ -677,19 +655,14 @@ namespace llarp
-          if (itr->second->PubKey() == pk)
-            endpoint = itr->second.get();
-        }
+        if (itr->second->PubKey() == pk)
```
this `if` seems redundant with the above changes.

tewinget replied:
actually it seems it may have been redundant before as well?

jagerman committed:
Remove debug comment

jagerman committed:
Abstract & simplify logic

majestrate commented on llarp/router/router.cpp:
```diff
@@ -69,13 +69,30 @@ namespace llarp
     _lastTick = llarp::time_now_ms();
     m_NextExploreAt = Clock_t::now();
+    m_Pump = _loop->make_waker([this]() {
+      paths.PumpDownstream();
```
move these two to PumpLL

jagerman committed:
Move paths pumping into PumpLL
